### ISSUE TITLE
Refactor Heroku deploy.py

### ISSUE THREAD

**Issue by ehmatthes** [ehmatthes/django-simple-deploy#275]
Once working, there's some refactoring to do.

First tasks
---

- [x] Run black to start.
- [x] Use `--json` wherever possible when parsing CLI calls. Start with `apps:info`.
- [x] Refactor deploy.py, based on experience of refactoring fly and simple_deploy.py.
- [x] Use a more consistent style for docstrings. See [PEP 257](https://peps.python.org/pep-0257/). See also `_get_deployed_project_name()` as an [example](https://github.com/ehmatthes/django-simple-deploy/blob/main/simple_deploy/management/commands/fly_io/deploy.py#L376).
- [x] Run black.
- [ ] Run tests.
    - [x] local
    - [ ] e2e
- [ ] Quick look at documentation
- [ ] Update CLI messages.
- [ ] Merge.

Second tasks
---

This is the last platform to be refactored. Go back through all three platforms, looking for utilities that can be pulled out into sd_utils. Any utility functions that can be put there can be much more reliably implemented and tested than leaving it to each platform to handle. This also creates much more consistent workflows across the different platforms that will be supported.

- [ ] Move remaining work to a new issue, and close this one.
- [ ] Use Parking Lot to close 5-10 oldest issues.
- [ ] Look at what's done across the different platform deploy scripts. Identify any utility functions that can be moved to sd_utils. Move them there, and make sure all work in the three deployer scripts is unique to that platform.
- [ ] Consider writing quick unit tests for each new utility function.

**Comment by ehmatthes** [comment]
Methods
---

- [ ] `__init__()`
- [ ] `deploy()`
- [ ] `_confirm_automate_all()`
- [ ] `_get_heroku_app_info()`
- [ ] `_set_heroku_env_var()`
- [ ] `_get_heroku_settings()`
- [ ] `_generate_procfile()`
- [ ] `_add_gunicorn()`
- [ ] `_check_allowed_hosts()`
- [ ] `_configure_db()`
- [ ] `_add_db_packages()`
- [ ] `_add_db_settings()`
- [ ] `_configure_static_files()`
- [ ] `_add_static_file_settings()`
- [ ] `_add_static_file_directory()`
- [ ] `_configure_debug()`
- [ ] `_configure_secret_key()`
- [ ] `_conclude_automate_all()`
- [ ] `_summarize_deployment()`
- [ ] `_show_success_message()`
- [ ] `_check_current_heroku_settings()`
- [ ] `_add_heroku_setting()`
- [ ] `_prep_heroku_setting()`
- [ ] `_generate_summary()`
- [ ] `_prep_automate_all()`
- [ ] `_validate_platform()`
- [ ] `handle_poetry()`

**Comment by ehmatthes** [comment]
More tasks
---

- [ ] Move `confirm_automate_all()` to simple_deploy.py? It doesn't seem to have any platform-specific behavior at this point, and it seems like it could be called right *before* calling the platform-specific deployer.

**Comment by ehmatthes** [comment]
`_validate_platform()`
---

- [ ] This method needs to verify a project was created, or we're using automate_all.

**Comment by ehmatthes** [comment]
Ongoing tasks
---

- [ ] When unit testing, `self.heroku_app_name` is not set.

**Comment by ehmatthes** [comment]
Simplify approach to settings
---

This is a much simpler approach to idempotency, and multiple runs. People will do multiple runs if something fails and they have to modify their project. Avoid that if at all possible, but keep our workflow as simple as possible.

- [ ] Rather than dealing with individual settings, just build a platform-specific settings block and add it at the end of the file.
- [ ] If we find a block with the same start line, ask if it's okay to overwrite that block. This should happen before making other changes.
    - [ ] Use `takewhile()` to keep only non-platform settings.
- [ ] See if sd_utils can offer a helper function for managing this for multiple platforms.
- [ ] Clarify suggestion for an unclean Git status.
- [ ] Implement for Fly.
- [ ] Implement for Platform.sh.
- [ ] Merge after doing this, to not diverge far from main when affecting other platforms.

**Comment by ehmatthes** [comment]
Simplify approach to requirements
---

- [ ] Similar to settings, don't add requirements one at a time. Keep a running list of requirements, and add them all as a block.
- [ ] Let simple_deploy pass responsibility for adding django-simple-deploy requirement to platform deployer. It would be nice if all simple_deploy does is add log, and modify .gitignore.
- [ ] Implement for Fly.
- [ ] Implement for Platform.sh.
- [ ] Merge.

**Comment by ehmatthes** [comment]
Simplify overall approach
---

Originally, I got caught up in replicating the configuration workflow that people do when following a platform's docs. There's no need to do that, and it's much more verbose than it needs to be.

The process really boils down to, and should look as simple as:

- [ ] Validate that the user's environment is set up to push to this platform, ie that the CLI is installed and authenticated, and any prerequisite resources have been created. This can include an app to deploy to, an env var for an API key, and other things that should be in place before any configuration changes are made.
- [ ] Set env vars on the platform.
- [ ] Add platform-specific requirements.
- [ ] Modify settings.

**Comment by ehmatthes** [comment]
Misc
---

- [ ] Looks like Platform.sh is confirming automate all twice, once in sd and once in its own script?

**Comment by ehmatthes** [comment]
Run e2e test
---

After first full pass refactoring heroku deploy.py, before porting some changes to other platforms.

- [ ] Run configuration-only e2e test.

Failed:

```sh
Migrating deployed project...
Running python manage.py migrate on ⬢ stark-coast-54964... up, run.2010 (Basic)
Traceback (most recent call last):
  File "/app/manage.py", line 22, in <module>
    main()
  File "/app/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/base.py", line 402, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/base.py", line 448, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/base.py", line 96, in wrapped
    res = handle_func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/core/management/commands/migrate.py", line 114, in handle
    executor = MigrationExecutor(connection, self.migration_progress_callback)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/migrations/executor.py", line 18, in __init__
    self.loader = MigrationLoader(self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/migrations/loader.py", line 58, in __init__
    self.build_graph()
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/migrations/loader.py", line 235, in build_graph
    self.applied_migrations = recorder.applied_migrations()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/migrations/recorder.py", line 81, in applied_migrations
    if self.has_table():
       ^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/migrations/recorder.py", line 57, in has_table
    with self.connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/backends/base/base.py", line 323, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/backends/base/base.py", line 299, in _cursor
    self.ensure_connection()
  File "/app/.heroku/python/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/backends/base/base.py", line 282, in ensure_connection
    self.connect()
  File "/app/.heroku/python/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/backends/base/base.py", line 262, in connect
    conn_params = self.get_connection_params()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/.heroku/python/lib/python3.12/site-packages/django/db/backends/postgresql/base.py", line 174, in get_connection_params
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: settings.DATABASES is improperly configured. Please supply the NAME or OPTIONS['service'] value.
  Opening project...
```

- [ ] Run automate-all e2e tests.
- [ ] Run pypi configuration-only e2e test.

**Comment by ehmatthes** [comment]
Make a new release
---

- [ ] Heroku works again, so go ahead and make a new release.

**Comment by ehmatthes** [comment]
All open tasks moved to #319.
